subsys/net/ip/net_mgmt.c

```diff
@@ -33,98 +33,65 @@ struct mgmt_event_wait {
 	struct net_if *iface;
 };
 
-static K_SEM_DEFINE(network_event, 0, K_SEM_MAX_LIMIT);
-static K_MUTEX_DEFINE(net_mgmt_lock);
+static K_MUTEX_DEFINE(net_mgmt_callback_lock);
+static K_MUTEX_DEFINE(net_mgmt_event_lock);
 
 K_KERNEL_STACK_DEFINE(mgmt_stack, CONFIG_NET_MGMT_EVENT_STACK_SIZE);
 static struct k_thread mgmt_thread_data;
-static struct mgmt_event_entry events[CONFIG_NET_MGMT_EVENT_QUEUE_SIZE];
 static uint32_t global_event_mask;
 static sys_slist_t event_callbacks = SYS_SLIST_STATIC_INIT(&event_callbacks);
-static int16_t in_event;
-static int16_t out_event;
+
+/* event structure used to prevent increasing the stack usage on the caller thread */
+static struct mgmt_event_entry new_event;
+K_MSGQ_DEFINE(event_msgq,
+	sizeof(struct mgmt_event_entry),
+	CONFIG_NET_MGMT_EVENT_QUEUE_SIZE,
+	sizeof(uint32_t));
+
 
 static inline void mgmt_push_event(uint32_t mgmt_event, struct net_if *iface,
 				   const void *info, size_t length)
 {
-	int16_t i_idx;
-
 #ifndef CONFIG_NET_MGMT_EVENT_INFO
 	ARG_UNUSED(info);
 	ARG_UNUSED(length);
 #endif /* CONFIG_NET_MGMT_EVENT_INFO */
 
-	(void)k_mutex_lock(&net_mgmt_lock, K_FOREVER);
+	(void)k_mutex_lock(&net_mgmt_event_lock, K_FOREVER);
 
-	i_idx = in_event + 1;
-	if (i_idx == CONFIG_NET_MGMT_EVENT_QUEUE_SIZE) {
-		i_idx = 0;
-	}
+	memset(&new_event, 0, sizeof(struct mgmt_event_entry));
 
 #ifdef CONFIG_NET_MGMT_EVENT_INFO
 	if (info && length) {
 		if (length <= NET_EVENT_INFO_MAX_SIZE) {
-			memcpy(events[i_idx].info, info, length);
-			events[i_idx].info_length = length;
+			memcpy(new_event.info, info, length);
+			new_event.info_length = length;
 		} else {
 			NET_ERR("Event %u info length %zu > max size %zu",
 				mgmt_event, length, NET_EVENT_INFO_MAX_SIZE);
-			(void)k_mutex_unlock(&net_mgmt_lock);
+			(void)k_mutex_unlock(&net_mgmt_event_lock);
 
 			return;
 		}
-	} else {
-		events[i_idx].info_length = 0;
 	}
 #endif /* CONFIG_NET_MGMT_EVENT_INFO */
 
-	events[i_idx].event = mgmt_event;
-	events[i_idx].iface = iface;
-
-	if (i_idx == out_event) {
-		uint16_t o_idx = out_event + 1;
-
-		if (o_idx == CONFIG_NET_MGMT_EVENT_QUEUE_SIZE) {
-			o_idx = 0U;
-		}
-
-		if (events[o_idx].event) {
-			out_event = o_idx;
-		}
-	} else if (out_event < 0) {
-		out_event = i_idx;
-	}
-
-	in_event = i_idx;
-
-	(void)k_mutex_unlock(&net_mgmt_lock);
-}
-
-static inline struct mgmt_event_entry *mgmt_pop_event(void)
-{
-	int16_t o_idx;
+	new_event.event = mgmt_event;
+	new_event.iface = iface;
 
-	if (out_event < 0 || !events[out_event].event) {
-		return NULL;
+	if (k_msgq_put(&event_msgq, &new_event, K_NO_WAIT)) {
```

Comment:
I wonder if introducing a small timeout here could improve the robustness of the net_mgmt even further. In case all slots are used, the function would block, giving chance the mgmt thread to process the events.

Reply:
Yeah, I've seen similar changes also, maybe a 10ms timeout won't harm. It's not that we're in need of realtime behavior.

Reply:
Added 10ms wait